---
title: "Design: top_genes()"
author: "Simona Rossomanno"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

## Objective

Update and replace the `draw_top_gene()` function from `rnaseqTools`. The function will be split in two functions: one function  `top_genes()`, to derive the top expressing genes and use an `HermesData` object as input, and a second function `draw_top_barplot()` to draw the bar plots using `ggplt2`.

## Idea
First we need to access the count data in Hermes data and select the top expressing ones. The selection could be based on the number of genes that we want to plot defined as `n_plot`, for example the top 10, or based on a given min average expression treshold `min_threshold` .

```{r}
#Creating the HermesData object from the example SummarizedExperiment object.
result <- HermesData(summarized_experiment)

#Accessing the counts data.
# for testing we could make sure that result is a hermesdata object 
count_results <- counts(result)

```
then we need to compute the average expression across the coulmns and order the data by the average expression.  

```{r}
# computing the average gene count across column
average_expression <- rowMeans(count_results)

```


We then subset based either on the first `n_top` or on `min_threshold` and create a dataframe. The dataframe will contain two columns and n rows (corrisponding to the number of rows in the counts matrix) containing one the name of the gene and the other the corrisponding average gene expression. The derived data has to be a dataframe because  `ggplots2` requires a data frame object. 

questions: 
    * is it better to have the data converted to dataframe in the `draw_top_barplot()` function instead of the `top_genes()` one?
    * do we need to plot nom,rlaised data instead?
    * should we account for more than one set? ei for an additional by group?

```{r}
#set as criteria number of genes for example the first 10
# here we could check that the numbers are positive integers
n_top <- 10L
min_threshold <- 50000L

# creating the subset dataframe 
library("dplyr")

average_expression <- data.frame(average_expression) %>%
  arrange(desc(average_expression)) 


``` 
the dataframe can then selected and  plotted. 

```{r}
# selecting on min_treshold
top_min_treshold <- average_expression %>%
   filter(average_expression >= min_threshold)

# or on n_top
top_n_top <- average_expression %>%
  slice_head(n = n_top , with_ties = TRUE)  #Should ties be kept together? The default, TRUE, may return more rows than you request. Use FALSE to ignore ties, and return the first n rows.


ggplot(top_min_treshold %>%
         arrange(desc(average_expression))) +
  geom_col (aes( y = average_expression, x = rownames(top_min_treshold))) +
  scale_x_discrete(name = "HGNC gene names") +
  scale_y_continuous(name = "Averaged CPM") 

# or 




``` 
QUESTION: what is the best option above with ties? I think if equaslly expressed whould be then reported all

So now putting the two functions togheter , need also to check which of the two option is going to be used
```{r}
top_genes <- function(object, 
                      n_top = 10L , 
                      min_threshold = FALSE) {
  
  
}

``` 

