---
params:
  
  title:
    label: "Title of the report [Customize title name as desired]"
    value: "RNASeq QC Report"
    input: text
  author: 
    label: "Author(s) of the report: <NAME> (Team/Department i.e. PDD DSX)"
    value: "Martha Mailman (PDD - Data Science Acceleration (DSX))"
    input: text
  input_data_object:
    label: "[REQUIRED] data object input."
    value: "/home/rstudio/NEST/hermes/data/summarized_experiment.rda"
    input: file
  Hermes_QC_output_dir:
    label: "[REQUIRED] Path to Hermes QC output directory (this is the only requried parameter of the report)"
    value: "/home/"  # "~/XX/XXX"
    input: file
  min_cpm:
    label: "min_cpm (non-negative `number`) minimum CPM for each gene within the sample."
    value: 1 # default 1
    input: numeric
  min_cpm_prop:
    label: "min_cpm_prop (`proportion`) minimum percentage of samples with acceptable CPM of certain gene for low expression flagging."
    value: 0.25
    input: numeric
  min_corr:
    label: "min_corr (`proportion`) minimum Pearson correlation coefficient of CPM between samples for technical failure flagging."
    value: 0.5
    input: numeric
  min_depth:
    label: "min_depth (non-negative `count` or `NULL`) minimum library depth for low depth flagging. If `NULL`, this will be calculated as the first quartile minus 1.5 times the inter-quartile range of the library size (depth) of all samples. (So anything below the usual lower boxplot whisker would be too low.)"
    value: NULL
    input: numeric
  top_gene_n_top: 
    label: "Filter for top expressed genes"
    value: 10 # default 10L by specifing NULL, alternatively 5 [if top_gene_n_top not NULL, top_gene_min_threshold must be NULL]
    input: numeric
  top_gene_min_threshold:
    label: "Filter for minimum expression level threshold"
    value: NULL # default specifing NULL, alternatively 50000 [if top_gene_min_threshold not NULL, top_gene_n_top must be NULL]
    input: numeric
  assay_name:
    label: "Indicating the name of the assay of interest, with possible options: counts, cpm, tpm, rpkm, voom."
    value: "counts"
    input: text
  cor_method:
    label: "Specifies the correlation method, see [stats::cor()] for details."
    value: "pearson"
    input: text
title: "`r params$title`"
author: '`r params$author`'
date: '`r format(Sys.time(), "%d %B %Y")`'
output:
  html_document:
    self_contained: true
    code_folding: hide
    mathjax: default
    number_sections: true
    theme: yeti
    toc: true
    toc_depth: 2
    toc_float:
      collapsed: false
      smooth_scroll: false
editor_options:
  chunk_output_type: console
---
```{r setup, message = FALSE}
# -----------------------------------------------------------------------------
# Developer

# The name and version of this report template (printed in last section)
#templateName <- params$title
templateName <- "qc_report"
templateVersion <- "1.0.0"
# Set TRUE to enable rmarkdown cache
knitr::opts_chunk$set(cache = FALSE, cache.lazy = FALSE)
# Set FALSE to skip all output (plots, tables), useful for debugging 
showOutput <- TRUE

# -----------------------------------------------------------------------------

# Knitr options (important to set error=FALSE, otherwise the evaluation does not stop after an error!)
knitr::opts_chunk$set(eval = TRUE, echo = TRUE, message = TRUE, warning = TRUE, error = FALSE)

# Libraries
library(hermes)
```

```{r initialize}
# Check that required parameters were set by user
#requiredPaths <- c("biokit_output_dir")
#checkPathsExist(params, requiredPaths)
#requiredParams <- c("color_vars")
#checkParamsNotEmpty(params, requiredParams)

# -----------------------------------------------------------------------------
```

```{r}
# Load Data and get object HermesData 
# Section 1 - Pre Gene Filtering
obj_name <- load(params$input_data_object)
se <- get(obj_name)
result <- HermesData(se)

# Section 2 - Post Gene Filtering & Normalization
control <- control_quality(params$min_cpm, params$min_cpm_prop, params$min_corr, params$min_depth)
result_fil <- HermesData(se) %>%
  add_quality_flags(control) %>%
  filter(h_low_expression_flag==0, h_low_depth_flag==0, h_tech_failure_flag==0) %>%
  normalize()
```

# Dataset Summary

The data set was composed of `r ncol(result)` samples and `r nrow(result)` genes. 

## Technical Metrics {.tabset}

### Histogram of Library Sizes 

First, we check how the distribution of library sizes looks like across all samples. The median library size is <b>`r sprintf("%0.3G", median(colSums(counts(result))))`</b>. 
According to [Standards, Guidelines and Best Practices for RNA-Seq (The ENCODE Consortium)](https://www.encodeproject.org/documents/cede0cbe-d324-4ce7-ace4-f0c3eddf5972/@@download/attachment/ENCODE%20Best%20Practices%20for%20RNA_v2.pdf "ENCODE Project Homepage"), the sequencing depth/library size is usually determined by the goals of the experiment and the nature of the RNA sample. The minimum library size required is around 20-30 million aligned reads. 

```{r genes_hist_libdepth,results='asis'}
draw_libsize_hist(result)
```

### Q-Q Plot of Library Sizes {.tabset}

The distribution of library sizes is then compared to the normal distribution. The linearity of the points will suggest if the distribution of library sizes is normally distributed.

```{r genes_qqplot_libdepth, results = 'asis'}
draw_libsize_qq(result)
```

### Density Plot of (Log) Counts Distributions {.tabset}

In this plot, each line represents density of expression of all the genes within one sample. Lines that deviate from the majority may suggest inconsistent read depth or technical failure of the samples. The peaks in log2 (count) distribution indicate the log2(count) value at which the majority of genes are expressed. 

```{r genes_density_libdepth, results = "asis"}
draw_libsize_densities(result)
```
## {-}


# Gene filtering {.tabset}

This step is necessary to flag out genes with low expression. Genes with very low counts across all libraries provide little evidence for differential expression and they interfere with some of the statistical analyses conducted later in the pipeline. Therefore, these genes should be filtered out prior to the analysis. 

There are many ways to filter out genes with lower counts. When there are n biological replicates in each group, we usually filter on a minimum counts per million threshold present in at least n samples. In this dataset, we choose to retain genes if they are expressed at a counts-per-million (CPM) above <b> `r params$min_cpm` </b> in at least in <b> `r (params$min_cpm_prop)*100`% </b> of the samples. These thresholds can be modified.

Total number of genes before filtering: `r nrow(result)` <br />
Total number of genes after filtering:  `r nrow(result_fil)` <br />
Rate of passing: `r round((100/nrow(result))*nrow(result_fil),1)` <br>

## Stacked Barplot of Low Expression Genes by Chromosome

This barplot shows the chromosomes with their proportions of low expression genes.

```{r}
draw_genes_barplot(result)
```

## Genes with extremely high counts
```{r}
result_tg <- top_genes(result, n_top = params$top_gene_n_top, min_threshold = params$top_gene_min_threshold)
```

This bar plot shows the shows the expression counts (y axis) for each of the <b> `r nrow(result_tg)` </b> top genes (x-axis) with.

```{r}
autoplot(result_tg, y_lab = "Maximum Count", title = "Top genes")
```

## Genes with high variability

This bar plot shows the shows the expression standard deviation (y axis) for each of the <b> `r nrow(result_tg)` </b> top genes (x-axis) with.

```{r}
result_sds <- top_genes(result, summary_fun = rowSds, n_top = params$top_gene_n_top, min_threshold = params$top_gene_min_threshold)
autoplot(result_sds, y_lab = "Standard Deviation across samples", title = "Top variability genes")
```
{-}

# Sample filtering {.tabset}

This step is necessary to flag out samples with low depth or technical failure. There are three approaches:   

## Technical failure

Human usually has similar RNAseq profile; therefore, correlation of RNAseq among patients should be high. If a patient's absolute correlation score is lower than a threshold (default = 0.5, <b>`r params$min_corr` </b> used in this report), it is likely that the sample preparation went wrong.

## Low depth samples 

Identified by lower quartile minus 1.5*IQR (default) or user-defined threshold: <b> `r params$top_gene_min_threshold` </b>Â 

## Boxplot of Non-zero Genes

In this plot, each data point represents the number of non-zero expressed genes in a sample. Samples with a lower number non-zero count genes may be also of interest. 

```{r}
draw_nonzero_boxplot(result) 
```

## Principal Components Analysis

Perform principal components analysis of the gene count vectors across all samples. PCA should be performed after filtering out low quality genes and samples and normalization. In addition, genes with constant counts across all samples are excluded from the analysis internally. Centering and scaling is applied internally. In this analysis the `r params$assay_name` assay was used.

```{r}
result_pca <- calc_pca(result, assay_name = params$assay_name)
autoplot(result_pca, label = TRUE)
```

## Correlation between Samples of HermesData

Correlation matrix between the sample vectors of counts from a specified assay, containing quality flags (`TechnicalFailureFlag`, `LowDepthFlag`) describing the original input samples.

```{r}
result_cor <- calc_cor(result, method = params$cor_method) 
autoplot(result_cor)
```
{-}




```{r, echo=FALSE}

# For reference rnaseqTool results
#gSE <- eset2gSE(example_ExpressionSet)
#gse.qc <- gQCproc(gSE) 
#gse.postqc <- subset(gse.qc, LowExpressionFlag == 0, LowDepthFlag == 0 & TechnicalFailureFlag == 0) # dropping low expression genes, samples with low depth or technical failure 
#getmeta(rowData(gse.postqc), "threshold.cpm")

```

<!-- ---------------------------------------------------------------------- -->
<!-- Prints the parameters and session information                          -->
<!-- Include at end of each biokitr report                                  -->
<!-- ---------------------------------------------------------------------- -->
```{r session_info, results = "asis"}
#printReproducibilitySection(params, templateName, templateVersion)   
```
