<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Introduction to HermesData()</title>


<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>






<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Introduction to <code>HermesData()</code></h1>
<h4 class="date">Last updated on 2021-05-25</h4>



<div id="the-hermesdata" class="section level2">
<h2>The <code>HermesData</code></h2>
<div id="comparison-of-summarizedexperiment-gsummarizedexperiment-and-hermesdata" class="section level3">
<h3>Comparison of <code>SummarizedExperiment</code>, <code>gSummarizedExperiment</code> and <code>HermesData</code></h3>
<div id="summarizedexperiment" class="section level4">
<h4><code>SummarizedExperiment</code></h4>
<p>The <code>SummarizedExperiment</code> class is similar to the historical <code>ExpressionSet</code>, the main distinction being that <code>SummarizedExperiment</code> is more flexible in its row information, allowing both <code>GRanges</code> based as well as those described by arbitrary DataFrames. This makes it ideally suited to a variety of experiments, particularly sequencing based experiments such as RNA-Seq and ChIP-Seq.</p>
</div>
<div id="gsummarizedexperiment" class="section level4">
<h4><code>gSummarizedExperiment</code></h4>
<p>The <code>gSummarizedExperiment</code> class in <code>rnaseqTools</code> is an extention to the <code>SummarizedExperiment</code> class by imposing mandatory columns on both the feature data (rowData) and the phenotype data (colData). So the structure of a <code>gSummarizedExperiment</code> object is exactly the same as a <code>SummarizedExperiment</code> object except that in a <code>gSummarizedExperiment</code> object:</p>
<ul>
<li><strong>rowData</strong> must contain the following variables that are retrived from the <code>igis</code> package:
<ul>
<li>HGNC: standardised nomenclature to human genes assigned by the The HUGO Gene Nomenclature Committee</li>
<li>HGNCGeneName: HGNC gene names</li>
<li>GeneID: Unique stable identifiers for genes and other loci for a subset of model organisms, using either EntrezID (by Entrez Gene) or EnsemblID (Ensembl databse).</li>
<li>Chromosome: Chromosome number of a given feature</li>
<li>StartBP: Genomic start position (as in base pair) of a given gene</li>
<li>EndBP: Genomic end position (as in base pair) of a given gene</li>
<li>WidthBP: length (as in base pair) of a given gene</li>
<li>CanonicalTransript:</li>
<li>ProteinTranscript:</li>
<li>LowExpressionFlag: whether a gene should be considered as a low expressed gene and exclude from drownstream analyses.</li>
</ul></li>
<li><strong>colData</strong> must contain variables:
<ul>
<li>Filename: path to the raw fastq files</li>
<li>SampleID: sample identification</li>
<li>LowDepthFlag: suggests whether a sample’s libarary size is smaller than usual and be excluded.</li>
<li>TechnicalFailureFlag: suggests whether a sample may be considered a technical failure and be excluded.</li>
</ul></li>
</ul>
</div>
<div id="hermesdata" class="section level4">
<h4><code>HermesData</code></h4>
<p>The <code>HermesData</code> object has similarities from the <code>SummarizedExperiment</code> and the <code>gSummarizedExperiment</code> objects. To be specific,</p>
<ul>
<li><p>Similar to the <code>gSummarizedExperiment()</code>, <code>HermesData()</code> must have mandatory rowData (HGNC, HGNCGeneName, etc) and colData (Filename, SampleID, etc).</p>
<ul>
<li>At this point, we don’t look up GeneID row data. It will be updated later.</li>
</ul></li>
<li><p>The <code>gSummarizedExperiment()</code> takes <code>assays</code>, <code>rowData</code>, <code>colData</code> and <code>metadata</code> as input, while <code>HermesData()</code> directly takes <code>SummarizedExperiment</code> object as input.</p></li>
</ul>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
